timestamp-impl/src/test/java/com/palantir/timestamp/PersistentUpperLimitTest.java

```diff
@@ -75,34 +76,50 @@ public void shouldOnlyMakeOneGetCallToTheUnderlyingStore() {
 
     @Test
     public void shouldIncreaseTheUpperLimitIfTheNewLimitIsBigger() {
-        upperLimit.increaseToAtLeast(TIMESTAMP);
+        upperLimit.increaseToAtLeast(TIMESTAMP, 0);
 
-        upperLimit.increaseToAtLeast(TIMESTAMP + 1000);
+        upperLimit.increaseToAtLeast(TIMESTAMP + 1000, 0);
         assertThat(upperLimit.get(), is(TIMESTAMP + 1000));
     }
 
     @Test
+    public void shouldIncreaseTheUpperLimitWithBufferIfTheNewLimitIsBigger() {
+        upperLimit.increaseToAtLeast(TIMESTAMP, 0);
+
+        upperLimit.increaseToAtLeast(TIMESTAMP + 1000, 1000);
+        assertThat(upperLimit.get(), is(TIMESTAMP + 2000));
+    }
+
+    @Test
     public void shouldNotIncreaseTheUpperLimitIfTheNewLimitIsSmaller() {
-        upperLimit.increaseToAtLeast(TIMESTAMP);
+        upperLimit.increaseToAtLeast(TIMESTAMP, 0);
 
-        upperLimit.increaseToAtLeast(TIMESTAMP - 1000);
+        upperLimit.increaseToAtLeast(TIMESTAMP - 1000, 0);
+        assertThat(upperLimit.get(), is(TIMESTAMP));
+    }
+
+    @Test
+    public void shouldNotIncreaseTheUpperLimitIfTheNewLimitIsSmallerRegardlessOfBuffer() {
+        upperLimit.increaseToAtLeast(TIMESTAMP, 0);
+
+        upperLimit.increaseToAtLeast(TIMESTAMP - 1000, 2000);
         assertThat(upperLimit.get(), is(TIMESTAMP));
     }
```

Comment:
Maybe one of these for equality as well? i.e. (TIMESTAMP, 0) then (TIMESTAMP, 1000) should still leave the upper limit as just TIMESTAMP.